include/ekuvector/ekuvector.hpp

```diff
@@ -0,0 +1,1122 @@
+typename ekuvector<Type, Allocator>::const_iterator
+ekuvector<Type, Allocator>::cbegin() const noexcept {
+  return data_;
+}
+
+/* *** */
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::end() noexcept {
+  return data_ + size_;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_iterator
+ekuvector<Type, Allocator>::end() const noexcept {
+  return data_ + size_;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_iterator
+ekuvector<Type, Allocator>::cend() const noexcept {
+  return data_ + size_;
+}
+
+/* *** */
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::reverse_iterator
+ekuvector<Type, Allocator>::rbegin() noexcept {
+  return reverse_iterator(end());
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_reverse_iterator
+ekuvector<Type, Allocator>::rbegin() const noexcept {
+  return const_reverse_iterator(end());
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_reverse_iterator
+ekuvector<Type, Allocator>::crbegin() const noexcept {
+  return const_reverse_iterator(end());
+}
+
+/* *** */
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::reverse_iterator
+ekuvector<Type, Allocator>::rend() noexcept {
+  return reverse_iterator{begin()};
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_reverse_iterator
+ekuvector<Type, Allocator>::rend() const noexcept {
+  return const_reverse_iterator{begin()};
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_reverse_iterator
+ekuvector<Type, Allocator>::crend() const noexcept {
+  return const_reverse_iterator{begin()};
+}
+
+/* *** */
+
+template <class Type, class Allocator>
+bool ekuvector<Type, Allocator>::empty() const noexcept {
+  return (size_ == 0);
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::size_type
+ekuvector<Type, Allocator>::size() const noexcept {
+  return size_;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::size_type
+ekuvector<Type, Allocator>::max_size() const noexcept {
+  return INT32_MAX;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::size_type
+ekuvector<Type, Allocator>::capacity() const noexcept {
+  return capacity_;
+}
+
+template <class Type, class Allocator>
+void ekuvector<Type, Allocator>::shrink_to_fit() {
+  /* the standard allows to ignore this request */
+}
+
+template <class Type, class Allocator>
+void ekuvector<Type, Allocator>::clear() noexcept {
+  while (size()) {
+    pop_back();
+  }
+}
+
+/* *** */
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::insert(const_iterator pos, const Type &value) {
```

Comment:
@glpuga both `ekuvector<Type, Allocator>::insert(const_iterator pos, const Type &value)` and `ekuvector<Type, Allocator>::insert(const_iterator pos, Type &&value)`  overloads implementations are somewhat inefficient, why not making both similar to that of `ekuvector<Type, Allocator>::insert(const_iterator pos, size_type count, const Type &value)` (i.e. moving a block, constructing in-place) ?

Reply:
I see you point. The **block-move-construct-in-place** algorithm requires me to move the preexisting elements one by one, making sure to to call the destructor on the evacuated place before doing anything else in that place to take care of the residual empty shell that moving leaves behind. All in all, if `N` elements need to be moved, it takes N move-constructs and N destructors + 1 copy/move construct for the final step

The **insert-last-and-bubble-down** version in the same case requires N move-constructs, 2N move-assign + N destructors + 1 copy/move construct for the initial insertion step.

So, yes, you're right, the **block-move-construct-in-place** requires less operations. The only reason for the algorithms to be different is that I implemented them at different times, and chose the slightly simpler insert-last-and-bubble-down for the first two, and opted for  **block-move-construct-in-place** for the more complex versions of insert by reusing the algorithm I used for `insert( const_iterator pos, InputIt first, InputIt last)`, which could not use the simpler algorithm (the elements inserted would be left in an undefined order).